### DISCUSSION TITLE
URL percent encode

### DISCUSSION THREAD

**Discussion by Indyandie** [curl/trurl#224] [Ideas]
> Follow up to this [comment](https://github.com/curl/trurl/discussions/219#discussioncomment-6235040)

Would it be useful to have the ability to percent encode parameter values in trurl?

For example, to encode the `test` parameter in the URL below. 

```url
http://test.com?test=%@#$ ^&*
```

Maybe something like this...

```shell
trurl "http://test.com?test=%@#$ ^&*" --percentencode "{query:test}"
```

output

```
http://test.com?test=%40%23%24%25%20%5E%26%2A
```

I tried using the `--urlencode` option but got an error due to the whitespace.

```shell
trurl "http://test.com?test=%@#$ ^&*" -g "{default:url}" --urlencode
```

```
trurl note: Malformed input to a URL function [http://test.com?test=%@#$ ^&*]
```

And after using the `--accept-space` it only encoded the whitespace character. 

```shell
trurl "http://test.com?test=%@#$ ^&*" -g "{default:url}" --urlencode --accept-space
```

```
http://test.com:80/?test=%@#$%20^&*
```

**Comment by emanuele6** [comment]
The URL was only rejected originally because because it contained a space, and you need `--accept-space` to make trurl accept spaces. 

If you have `?foo=bar#baz` then the query's `foo` parameter is `bar`. `baz` is not part of the query, it is the fragment of the URL.

In fact, note that if you try to normalise that URL with trurl, you will see that the space gets encoded as `%20`, and not `+`, because it is not part of the query:
```
$ trurl --accept-space "http://test.com?test=%@#$ ^&*"
http://test.com/?test=%@#$%20^&*
```
Removing the `#`, it becomes a `+` because it is actually part of the query:
```
$ trurl --accept-space "http://test.com?test=%@$ ^&*" 
http://test.com/?test=%@$+^&*
```

I do notice that `%` has not been encoded, which is not great, but I think that is caused by how libcurl handles URLs internally, and would need to be fixed there.

---

> `http://test.com?test=%40%23%24%25%20%5E%26%2A`

That is not an equivalent URL though; this URL has a query with only a `test` parameter equal to `%@$ ^&*`; the original URL had a query with a `test` parameter equal to `%@`, and a fragment equal to `$ ^&*`. Or, if we ignore the `#`, this URL has a query with only a `test` parameter equal to `%@$ ^&*`, while the original URL had a query with a `test` parameter equal to `%@$^` and a `*` parameter equal to *empty*. If you encode `&`s and/or `#`, you change the meaning of the URL.

I am trying to make sense of what you are trying to do here.
Are you perhaps writing a shell script in which you have a variable that contains a value, and you want to build a URL for `test.com` that contain `test=thatvalue` in the query, and you tried to do something like this?
```sh
var='foo%bar@$ ^&*'
trurl "https://test.com/?test=$var"
```

If so, that is not how you are supposed to use `trurl`, you should do something like this instead:
```sh
var='foo%7bar@$ ^&*'
trurl -a "query=test=$var" https://test.com/
# output: https://test.com/?test=foo%257bar%40%24%20%5e%26%2a
```
This will reliably set `test` in the query to the value `$var` whatever characters it contains, and without misinterpreting `%7b` as `{`.

Hopefully that helps.

o/
&nbsp;emanuele6
